ledger/participant-integration-api/src/main/scala/platform/store/state/StateCache.scala

```diff
@@ -0,0 +1,122 @@
+package com.daml.platform.store.state
+
+import java.util.concurrent.atomic.{AtomicLong, AtomicReference}
+
+import com.daml.caching.Cache
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.platform.store.state.StateCache.PendingUpdates
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+trait StateCache[K, V] {
+  protected implicit def ec: ExecutionContext
+  protected def cache: Cache[K, V]
+
+  protected val logger: ContextualizedLogger = ContextualizedLogger.get(getClass)
+
+  private[store] val pendingUpdates =
+    collection.concurrent.TrieMap.empty[K, PendingUpdates]
+
+  def get(key: K)(implicit loggingContext: LoggingContext): Option[V] =
+    cache.getIfPresent(key) match {
+      case Some(value) =>
+        logger.debug(s"Cache hit for $key -> ${value.toString.take(100)}")
+        Some(value)
+      case None =>
+        logger.debug(s"Cache miss for $key ")
+        None
+    }
+
+  def put(key: K, validAt: Long, value: V)(implicit
+      loggingContext: LoggingContext
+  ): Unit = putAsync(key, validAt, Future.successful(value))
+
+  def putAsync(key: K, validAt: Long, eventualValue: Future[V])(implicit
+      loggingContext: LoggingContext
+  ): Unit = {
+    logger.debug(s"New pending cache update for key $key at $validAt")
+    val pendingUpdate @ PendingUpdates(pendingCountRef, latestRef, effectsChain) =
+      pendingUpdates.getOrElseUpdate(key, PendingUpdates.empty)
+
+    /* We need to synchronize here instead of using a lock-free update
+     * since enqueueing the next update in the chain is impure
+     * and cannot be retried/cancelled on thread contention.
+     */
+    val needsRetry = pendingUpdate.synchronized {
+      /* Invalidated by a more recent put or a more recent pending update */
+      if (latestRef.get() >= validAt) false
+      /* Is this the first update in this effect chain? */
+      else if (latestRef.get() == Long.MinValue) {
+        pendingCountRef.set(1L)
+        latestRef.set(validAt)
+        effectsChain
+          .updateAndGet(_.transformWith { _ =>
+            registerEventualCacheUpdate(key, eventualValue, validAt).map(_ => ())
+          })
+        false
+      } else if (othersPending(pendingCountRef, latestRef)) {
+        /* If 0, another update is going to delete this entry, go for a retry. */
+        if (pendingCountRef.getAndIncrement() > 0L) {
+          latestRef.set(validAt)
+          effectsChain.updateAndGet(_.transformWith { _ =>
+            registerEventualCacheUpdate(key, eventualValue, validAt).map(_ => ())
+          })
+          false
+        } else latestRef.get() >= validAt
+      } else true
+    }
+    if (needsRetry) {
```

Comment:
My understanding is that the retry is required, as acquiring the `pendingUpdate` on which to synchronize is racy. Add commentary explaining this in detail to readers.

Reply:
Reminder from our discussion: ensure that the function is tail-recursive; and has a maximum retry boundary